LEGO1/mxpalette.cpp

```diff
@@ -1,21 +1,128 @@
 #include "mxpalette.h"
 
+// OFFSET: LEGO1 0x100bee30
+MxPalette::MxPalette()
+{
+  this->m_overrideSkyColor = FALSE;
+  this->m_palette = NULL;
+  GetDefaultPalette(this->m_entries);
+  this->m_skyColor = this->m_entries[141];
+}
+
+// OFFSET: LEGO1 100bef90
+MxPalette::~MxPalette()
+{
+  if (m_palette) {
+    m_palette->Release();
+  }
+}
+
+// OFFSET: LEGO1 0x100bf390
+void MxPalette::ApplySystemEntriesToPalette(LPPALETTEENTRY p_entries)
+{
+  // FIXME: incomplete
+  HDC hdc = GetDC(NULL);
+  int rastercaps = GetDeviceCaps(hdc, RASTERCAPS);
+  int sizepalettecaps;
+  if ((rastercaps & RC_PALETTE) != 0) {
+    sizepalettecaps = GetDeviceCaps(hdc, SIZEPALETTE);
+    if(sizepalettecaps = 256) {
+      GetSystemPaletteEntries(hdc,0,10,p_entries);
+      GetSystemPaletteEntries(hdc,246,10,p_entries + 0xf6);
+      ReleaseDC(NULL, hdc);
+    }
+  }
+  // FIXME: we get g_defaultPalette here, we need to define that, then we cna do the memcpy's
+}
+
+// OFFSET: LEGO1 100bf0b0
+MxPalette* MxPalette::Clone()
+{
+  // FIXME: doesnt match
+  MxPalette *pal = (MxPalette *) malloc(0x414);
+  if(pal != NULL) {
+    GetEntries(pal->m_entries);
+    pal->m_overrideSkyColor = m_overrideSkyColor;
+  }
+  return pal;
+}
+
+// OFFSET: LEGO1 0x100beed0
+MxPalette* MxPalette::FromBitmapPalette(RGBQUAD* p_bmp)
```

Comment:
'strings' doesn't return a 'FromBitmapPalette', so yeah, we can't say with 100% certainty that it is a ctor and not a method. It's likely that this is just another constructor but if it was, wouldn't it be visible in the dllspec?

Reply:
> 'strings' doesn't return a 'FromBitmapPalette', so yeah, we can't say with 100% certainty that it is a ctor and not a method. It's likely that this is just another constructor but if it was, wouldn't it be visible in the dllspec?
> 
> anyways: this requires ApplySystemEntriesToPalette

I can say with 100% certainty that it is a constructor.

Reply:
To perhaps help with the understanding that it is, with 100% certainty, a constructor:

1. How is it called? In this case, the caller allocates the memory with operator new, the size of the memory is equal to the size of the class, the memory is not initialized. The memory is transported in `ecx`. That makes it _certain_ it's either `__fastcall` or `__thiscall`. It looks like it's taking an argument on the stack, but at this point we can not yet know with 100% certainty whether or not that `push` is really for the immediately called function or if the call is interleaved, and the pushed argument is for a later function call.

2. How does it behave. First we look at the return instruction. It's `ret 4`. This verifies two things; that it is either a `__fastcall` with a stack-spilled argument or a `__thiscall`. Since `edx` is not set by the caller, and not used by the callee, it's with 100% certainty a `__thiscall` taking one argument. We then look at how the `this` memory is intitialized (we already know it's initialization taking place, since the memory is pristine). The code sets a vfptr. That verifies with 100% certainty that it must be either a ctor or a dtor. Since this method takes an argument, it's certain it is a ctor. However, there is one last piece of evidence verifying it is a ctor, that applies whether or not it takes any arguments: it returns `this`.

Armed with this knowledge, you can now with 100% certainty identify if a method is a constructor or a named method.

- Uninitialized memory in ecx. Check.
- Memory is exactly the size of the object (or at least larger, in case a subclass of this is inlined by the caller, following this call). Check.
- Method initializes and returns the memory. Check.
- Method sets the vfptr. Definitely Check. Only (!) constructors and destructors set the vfptr.
- Method returns `this`. Check.

(please note these are not the _only_ ways it can be verified, or that the generated ctor code can manifest itself - it can for example also be fully or partially inlined, i.e. directly following the call to operator new and the memory null check)


> anyways: this requires ApplySystemEntriesToPalette

That's easy. Only lacking the global default palette entries, yes?

```
static PALETTEENTRY g_10102188_rg256_DefaultPalette[256] =
{
  { 0, 0, 0, 0 },
  { 128, 0, 0, 0 },
  { 0, 128, 0, 0 },
  { 128, 128, 0, 0 },
  { 0, 0, 128, 0 },
  { 128, 0, 128, 0 },
  { 0, 128, 128, 0 },
  { 128, 128, 128, 0 },
  { 192, 220, 192, 0 },
  { 166, 202, 240, 0 },
  { 255, 255, 255, 0 },
  { 250, 250, 250, 0 },
  { 239, 239, 239, 0 },
  { 228, 228, 228, 0 },
  { 217, 217, 217, 0 },
  { 206, 206, 206, 0 },
  { 195, 195, 195, 0 },
  { 185, 185, 185, 0 },
  { 174, 174, 174, 0 },
  { 163, 163, 163, 0 },
  { 152, 152, 152, 0 },
  { 141, 141, 141, 0 },
  { 130, 130, 130, 0 },
  { 123, 123, 123, 0 },
  { 115, 115, 115, 0 },
  { 108, 108, 108, 0 },
  { 101, 101, 101, 0 },
  { 93, 93, 93, 0 },
  { 86, 86, 86, 0 },
  { 79, 79, 79, 0 },
  { 71, 71, 71, 0 },
  { 64, 64, 64, 0 },
  { 54, 54, 54, 0 },
  { 43, 43, 43, 0 },
  { 33, 33, 33, 0 },
  { 22, 22, 22, 0 },
  { 12, 12, 12, 0 },
  { 8, 8, 8, 0 },
  { 4, 4, 4, 0 },
  { 0, 0, 0, 0 },
  { 225, 218, 217, 0 },
  { 195, 182, 179, 0 },
  { 165, 145, 141, 0 },
  { 134, 108, 102, 0 },
  { 104, 72, 64, 0 },
  { 74, 35, 26, 0 },
  { 59, 28, 21, 0 },
  { 44, 21, 16, 0 },
  { 30, 14, 10, 0 },
  { 15, 7, 5, 0 },
  { 250, 231, 232, 0 },
  { 240, 185, 189, 0 },
  { 233, 154, 160, 0 },
  { 226, 124, 131, 0 },
  { 219, 93, 102, 0 },
  { 213, 62, 73, 0 },
  { 203, 18, 32, 0 },
  { 172, 15, 27, 0 },
  { 159, 14, 25, 0 },
  { 146, 13, 23, 0 },
  { 133, 12, 21, 0 },
  { 120, 11, 19, 0 },
  { 107, 10, 17, 0 },
  { 94, 8, 15, 0 },
  { 81, 7, 13, 0 },
  { 68, 6, 11, 0 },
  { 55, 5, 9, 0 },
  { 42, 4, 7, 0 },
  { 29, 3, 5, 0 },
  { 10, 1, 2, 0 },
  { 227, 236, 242, 0 },
  { 178, 203, 220, 0 },
  { 145, 181, 205, 0 },
  { 112, 159, 191, 0 },
  { 79, 137, 176, 0 },
  { 30, 104, 154, 0 },
  { 0, 84, 140, 0 },
  { 0, 79, 132, 0 },
  { 0, 72, 119, 0 },
  { 0, 66, 110, 0 },
  { 0, 61, 101, 0 },
  { 0, 55, 92, 0 },
  { 0, 47, 78, 0 },
  { 0, 39, 65, 0 },
  { 0, 34, 56, 0 },
  { 0, 28, 47, 0 },
  { 0, 23, 38, 0 },
  { 0, 18, 29, 0 },
  { 0, 12, 20, 0 },
  { 0, 4, 7, 0 },
  { 230, 242, 234, 0 },
  { 180, 215, 193, 0 },
  { 147, 198, 166, 0 },
  { 113, 180, 138, 0 },
  { 80, 162, 111, 0 },
  { 30, 136, 70, 0 },
  { 0, 120, 45, 0 },
  { 0, 114, 43, 0 },
  { 0, 102, 38, 0 },
  { 0, 95, 35, 0 },
  { 0, 83, 31, 0 },
  { 0, 72, 27, 0 },
  { 0, 63, 24, 0 },
  { 0, 56, 21, 0 },
  { 0, 48, 18, 0 },
  { 0, 36, 14, 0 },
  { 0, 25, 9, 0 },
  { 0, 17, 6, 0 },
  { 0, 9, 3, 0 },
  { 0, 1, 1, 0 },
  { 254, 244, 220, 0 },
  { 255, 239, 181, 0 },
  { 255, 231, 156, 0 },
  { 255, 222, 132, 0 },
  { 255, 222, 115, 0 },
  { 255, 214, 99, 0 },
  { 255, 206, 66, 0 },
  { 255, 198, 41, 0 },
  { 255, 185, 0, 0 },
  { 255, 189, 8, 0 },
  { 247, 181, 0, 0 },
  { 222, 156, 0, 0 },
  { 189, 140, 0, 0 },
  { 173, 123, 0, 0 },
  { 148, 107, 0, 0 },
  { 132, 90, 0, 0 },
  { 107, 74, 0, 0 },
  { 74, 49, 0, 0 },
  { 57, 41, 0, 0 },
  { 33, 24, 0, 0 },
  { 117, 52, 87, 0 },
  { 176, 158, 50, 0 },
  { 122, 165, 29, 0 },
  { 242, 142, 8, 0 },
  { 164, 43, 36, 0 },
  { 113, 67, 20, 0 },
  { 255, 0, 255, 0 },
  { 255, 0, 255, 0 },
  { 255, 0, 255, 0 },
  { 255, 0, 255, 0 },
  { 255, 0, 255, 0 },
  { 57, 163, 217, 0 },
  { 255, 255, 255, 0 },
  { 254, 255, 247, 0 },
  { 253, 253, 239, 0 },
  { 248, 247, 247, 0 },
  { 248, 247, 231, 0 },
  { 240, 240, 240, 0 },
  { 239, 239, 218, 0 },
  { 227, 232, 236, 0 },
  { 224, 221, 209, 0 },
  { 215, 222, 215, 0 },
  { 213, 214, 215, 0 },
  { 214, 214, 203, 0 },
  { 255, 219, 57, 0 },
  { 206, 206, 206, 0 },
  { 206, 206, 198, 0 },
  { 255, 214, 18, 0 },
  { 207, 203, 186, 0 },
  { 197, 199, 199, 0 },
  { 255, 206, 0, 0 },
  { 207, 198, 159, 0 },
  { 247, 204, 0, 0 },
  { 189, 198, 189, 0 },
  { 189, 189, 189, 0 },
  { 238, 199, 0, 0 },
  { 189, 189, 181, 0 },
  { 238, 190, 24, 0 },
  { 181, 189, 184, 0 },
  { 161, 186, 224, 0 },
  { 181, 181, 181, 0 },
  { 231, 189, 0, 0 },
  { 173, 182, 173, 0 },
  { 222, 181, 0, 0 },
  { 173, 173, 173, 0 },
  { 213, 182, 0, 0 },
  { 172, 173, 160, 0 },
  { 214, 173, 0, 0 },
  { 165, 165, 165, 0 },
  { 206, 173, 0, 0 },
  { 160, 168, 151, 0 },
  { 206, 164, 0, 0 },
  { 198, 165, 0, 0 },
  { 157, 156, 156, 0 },
  { 134, 156, 200, 0 },
  { 153, 156, 144, 0 },
  { 142, 156, 161, 0 },
  { 189, 156, 0, 0 },
  { 148, 148, 148, 0 },
  { 146, 148, 138, 0 },
  { 133, 143, 161, 0 },
  { 189, 143, 0, 0 },
  { 140, 140, 140, 0 },
  { 177, 147, 0, 0 },
  { 131, 140, 136, 0 },
  { 146, 130, 126, 0 },
  { 170, 137, 0, 0 },
  { 132, 132, 130, 0 },
  { 123, 125, 125, 0 },
  { 123, 123, 133, 0 },
  { 153, 126, 0, 0 },
  { 114, 116, 118, 0 },
  { 110, 112, 108, 0 },
  { 97, 109, 136, 0 },
  { 127, 108, 6, 0 },
  { 0, 173, 0, 0 },
  { 100, 99, 101, 0 },
  { 176, 71, 41, 0 },
  { 36, 142, 33, 0 },
  { 98, 91, 75, 0 },
  { 80, 88, 104, 0 },
  { 252, 0, 0, 0 },
  { 78, 71, 73, 0 },
  { 73, 71, 78, 0 },
  { 62, 63, 61, 0 },
  { 0, 66, 211, 0 },
  { 99, 51, 14, 0 },
  { 198, 0, 0, 0 },
  { 189, 0, 0, 0 },
  { 0, 57, 206, 0 },
  { 181, 0, 0, 0 },
  { 0, 56, 185, 0 },
  { 173, 0, 0, 0 },
  { 165, 0, 0, 0 },
  { 49, 49, 49, 0 },
  { 0, 49, 165, 0 },
  { 156, 0, 0, 0 },
  { 42, 45, 60, 0 },
  { 148, 0, 0, 0 },
  { 140, 0, 0, 0 },
  { 41, 41, 41, 0 },
  { 0, 41, 144, 0 },
  { 132, 0, 0, 0 },
  { 123, 0, 0, 0 },
  { 7, 35, 114, 0 },
  { 34, 36, 32, 0 },
  { 115, 0, 0, 0 },
  { 107, 0, 0, 0 },
  { 90, 0, 0, 0 },
  { 23, 24, 27, 0 },
  { 74, 0, 0, 0 },
  { 15, 15, 16, 0 },
  { 49, 0, 0, 0 },
  { 16, 12, 4, 0 },
  { 7, 8, 8, 0 },
  { 0, 0, 8, 0 },
  { 255, 251, 240, 0 },
  { 160, 160, 164, 0 },
  { 128, 128, 128, 0 },
  { 255, 0, 0, 0 },
  { 0, 255, 0, 0 },
  { 255, 255, 0, 0 },
  { 0, 0, 255, 0 },
  { 255, 0, 255, 0 },
  { 0, 255, 255, 0 },
  { 255, 255, 255, 0 }
};
```

Add that array to mxpalette.cpp, and the copy code to both `ApplySystemEntriesToPalette` and `GetDefaultPalette` pretty much write itself (not literally, obviously).